fix(ngcc): support recovering when a worker process crashes (and more)

gkalpak opened the pull request:
Several small ngcc fixes; the main one being being able to recover when a worker process crashes.
See individual commit messages for more info.

Jira issue: [FW-2008](https://angular-team.atlassian.net/browse/FW-2008)

Fixes #36278.

mary-poppins commented:
You can preview cc6daf8 at https://pr36626-cc6daf8.ngbuilds.io/.

mary-poppins commented:
You can preview 8f92a83 at https://pr36626-8f92a83.ngbuilds.io/.

mary-poppins commented:
You can preview ed19d80 at https://pr36626-ed19d80.ngbuilds.io/.

mary-poppins commented:
You can preview 8335429 at https://pr36626-8335429.ngbuilds.io/.

mary-poppins commented:
You can preview 777cf8f at https://pr36626-777cf8f.ngbuilds.io/.

mary-poppins commented:
You can preview 17b4fac at https://pr36626-17b4fac.ngbuilds.io/.

mary-poppins commented:
You can preview 1f3257b at https://pr36626-1f3257b.ngbuilds.io/.

mary-poppins commented:
You can preview 03e74f2 at https://pr36626-03e74f2.ngbuilds.io/.

petebacondarwin commented:
One can argue that none of these commits are "feat" since they don't actually offer any new features to the end-user. Instead they are parts of a reorganisation (refactor) of the ngcc internal implementation to "fix" the problem of a lack of resilience to workers being killed.

(And features can't be cherry picked to stable).

petebacondarwin commented:
typo in `fix(ngcc): give up re-spawing crashed worker process after 3 attempts`:

spawing -> spawning

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/in_place_file_writer_spec.ts:
```diff
@@ -29,77 +29,140 @@ runInEachFileSystem(() => {
+        expect(fs.exists(fileBackupPath)).toBeFalse();
+      });
+
+      it('should only revert the written file if there is no backup', () => {
```
This test description is a bit misleading, since we do revert the file (to nothing). The implication of not reverting would be to leave the file untouched, which is not what happens. Perhaps instead it should be something like:

```suggestion
      it('should just remove the written file if there is no backup', () => {
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/new_entry_point_file_writer_spec.ts:
```diff
@@ -494,6 +493,82 @@ runInEachFileSystem(() => {
+        esm5bundle = makeTestBundle(fs, entryPoint, 'module', 'esm5');
+      });
+
+      it('should revert a written JS file', () => {
```
```suggestion
      it('should remove non-typings files', () => {
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/writing/new_entry_point_file_writer.ts:
```diff
@@ -45,6 +45,17 @@ export class NewEntryPointFileWriter extends InPlaceFileWriter {
+      this.fs.removeFile(newFilePath);
+    }
+  }
+
```
What about reverting all the files that are copied across in `copyBundle()`?

petebacondarwin commented on packages/compiler-cli/ngcc/src/execution/cluster/worker.ts:
```diff
@@ -28,24 +28,23 @@ if (require.main === module) {
       const {
-        createNewEntryPointFormats = false,
-        logger = new ConsoleLogger(LogLevel.info),
+        logger,
```
Good spot!

petebacondarwin commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -95,19 +96,20 @@ export function mainNgcc(options: NgccOptions): void|Promise<void> {
       propertiesToConsider, inParallel);
 
-  // Create an updater that will actually write to disk. In
+  // Create an updater that will actually write to disk.
```
👍

petebacondarwin reviewed (changes requested):
Great PR @gkalpak - the commits are really clean and easy to follow.
There are a bunch of nits but the two main things I think we should address are:

1) How to revert the `copyBundle()` action in the `NewEntryPointFileWriter`
  One could argue that we don't need to bother reverting these files since they will be overwritten automatically when the task is re-run. But perhaps we ought to be explicit about this in the code in case something changes in who `writeBundle()` works.

2) What happens if the process dies after sending `update-package-json` but before sending `task-completed`.
  I think we will need a way to undo the changes to `package.json` in this case. Alternatively we could defer actually writing to package.json until the `task-completed` message arrives.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/tasks/queues/base_task_queue.ts:
```diff
@@ -63,6 +63,16 @@ export abstract class BaseTaskQueue implements TaskQueue {
+    }
+
+    this.inProgressTasks.delete(task);
+    this.tasks.unshift(task);
```
Exactly.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/worker.ts:
```diff
@@ -81,10 +81,21 @@ export async function startWorker(logger: Logger, createCompileFn: CreateCompile
+          //
+          // Exit with an error and let the cluster master decide how to handle this.
+          logger.warn(`[Worker #${cluster.worker.id}] ${err.stack || err.message}`);
+          return process.exit(1);
```
Yes :grin:

gkalpak commented on packages/compiler-cli/ngcc/src/execution/tasks/api.ts:
```diff
@@ -115,7 +115,7 @@ export interface TaskQueue {
    * Do not process the tasks that depend upon the given task.
    */
-  markAsFailed(task: Task): void;
+  markTaskFailed(task: Task): void;
```
I don't feel strongly about it. Happy to go with `markAs*` (as long as the names are consistent :grin:).

gkalpak commented:
I added a bunch of fixup commits. PTAL.

As discussed "offline", we will not be reverting `copyBundle()` changes for now. (Added a comment to explain that.)
I also added code to revert the `updatePackageJson()` changes. For simplicity, because `updatePackageJson()` only adds properties (i.e. it does not modify existing ones), we can just remove the added properties (if they exist).

mary-poppins commented:
You can preview 7924a80 at https://pr36626-7924a80.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/ngcc/src/writing/new_entry_point_file_writer.ts:
```diff
@@ -45,15 +45,25 @@ export class NewEntryPointFileWriter extends InPlaceFileWriter {
+    // are identical to the original ones and they will be overwritten when re-processing the
+    // entry-point anyway.
+    //
+    // This way, we avoid the overhed of having to inform the master process about all source files
```
```suggestion
    // This way, we avoid the overhead of having to inform the master process about all source files
```

mary-poppins commented:
You can preview f1af402 at https://pr36626-f1af402.ngbuilds.io/.

mary-poppins commented:
You can preview 932a772 at https://pr36626-932a772.ngbuilds.io/.

gkalpak commented:
**merge-assistance:** The `size` check failure is unrelated to this PR. It looks like a CI/infrastructure issue.

gkalpak committed:
refactor(ngcc): move "Compiling" log message before starting work on a task

Previously, the "Compiling <entryPoint>" log message was printed before
starting to analyze and transform files, but after creating the
`EntryPointBundle` (which includes creating the TS program).

Since creating the `EntryPointBundle` involves some work, it is more
accurate to move the log message before creating the bundle.

gkalpak committed:
fix(ngcc): do not run in parallel mode if there are less than 3 CPU cores

Previously, ngcc would run in parallel mode (using the
`ClusterExecutor`) when there were at least 2 CPU cores (and all other
requirements where met). On systems with just 2 CPU cores, this meant
there would only be one worker process (since one CPU core is always
reserved for the master process). In these cases, the tasks would still
be processed serially (on the one worker process), but we would also pay
the overhead of communicating between the master and worker processes.

This commit fixes this by only running in parallel mode if there are
more than 2 CPU cores (i.e. at least 2 worker processes).

gkalpak committed:
feat(ngcc): support marking an in-progress task as unprocessed

This commit adds support for stopping processing an in-progress task
and moving it back to the list of pending tasks.

In a subsequent commit, this will be used to allow ngcc to recover when
a worker process crashes in the middle of processing a task.

gkalpak committed:
fixup! feat(ngcc): support marking an in-progress task as unprocessed

gkalpak committed:
refactor(ngcc): rename `TaskQueue#markTaskCompleted()` to `markAsCompleted()`

Rename the `markTaskCompleted()` method to be consistent with the other
similar methods of `TaskQueue` (`markAsFailed()` and
`markAsUnprocessed()`).

gkalpak committed:
refactor(ngcc): support running callback before writing transformed files

This commit enhances the `CompileFn`, which is used to process each
entry-point, to support running a passed-in callback (and wait for it to
complete) before proceeding with writing the transformed files to disk.

This functionality is currently not used. In a subsequent commit, it
will be used for passing info from worker processes to the master
process that will allow ngcc to recover when a worker process crashes in
the middle of processing a task.

gkalpak committed:
refactor(ngcc): notify master process about transformed files before writing

With this commit, worker processes will notify the master process about
the transformed files they are about to write to disk before starting
writing them.

In a subsequent commit, this will be used to allow ngcc to recover when
a worker process crashes in the middle of processing a task.

gkalpak committed:
fixup! refactor(ngcc): notify master process about transformed files before writing

gkalpak committed:
refactor(ngcc): keep track of transformed files per task

With this commit, the master process will keep track of the transformed
files that each worker process is intending to write to disk.

In a subsequent commit, this info will be used to allow ngcc to recover
when a worker process crashes in the middle of processing a task.

gkalpak committed:
feat(ngcc): support reverting a file written by `FileWriter`

This commit adds a `revertFile()` method to `FileWriter`, which can
revert a transformed file (and its backup - if any) written by the
`FileWriter`.

In a subsequent commit, this will be used to allow ngcc to recover
when a worker process crashes in the middle of processing a task.

gkalpak committed:
fixup! feat(ngcc): support reverting a file written by `FileWriter`

gkalpak committed:
fixup! feat(ngcc): support reverting a file written by `FileWriter`

gkalpak committed:
fix(ngcc): support recovering when a worker process crashes

Previously, when running in parallel mode and a worker process crashed
while processing a task, it was not possible for ngcc to continue
without risking ending up with a corrupted entry-point and therefore it
exited with an error. This, for example, could happen when a worker
process received a `SIGKILL` signal, which was frequently observed in CI
environments. This was probably the result of Docker killing processes
due to increased memory pressure.

One factor that amplifies the problem under Docker (which is often used
in CI) is that it is not possible to distinguish between the available
CPU cores on the host machine and the ones made available to Docker
containers, thus resulting in ngcc spawning too many worker processes.

This commit addresses these issues in the following ways:

1. We take advantage of the fact that files are written to disk only
   after an entry-point has been fully analyzed/compiled. The master
   process can now determine whether a worker process has not yet
   started writing files to disk (even if it was in the middle of
   processing a task) and just put the task back into the tasks queue if
   the worker process crashes.

2. The master process keeps track of the transformed files that a worker
   process will attempt to write to disk. If the worker process crashes
   while writing files, the master process can revert any changes and
   put the task back into the tasks queue (without risking corruption).

3. When a worker process crashes while processing a task (which can be a
   result of increased memory pressure or too many worker processes),
   the master process will not try to re-spawn it. This way the number
   or worker processes is gradually adjusted to a level that can be
   accomodated by the system's resources.

Examples of ngcc being able to recover after a worker process crashed:
- While idling: https://circleci.com/gh/angular/angular/682197
- While compiling: https://circleci.com/gh/angular/angular/682209
- While writing files: https://circleci.com/gh/angular/angular/682267

Jira issue: [FW-2008](https://angular-team.atlassian.net/browse/FW-2008)

Fixes #36278

gkalpak committed:
fixup! fix(ngcc): support recovering when a worker process crashes

gkalpak committed:
fix(ngcc): give up re-spawing crashed worker process after 3 attempts

Previously, when the last worker process crashed, the master process
would try to re-spawn it indefinitely. This could lead to an infinite
loop (if for some reason the worker process kept crashing).

This commit avoids this by limiting the number of re-spawn attempts to
3, after which ngcc will exit with an error.

gkalpak committed:
fix(ngcc): handle `ENOMEM` errors in worker processes

When running in parallel mode, worker processes forward errors thrown
during task processing to the master process, which in turn exits with
an error.

However, there are cases where the error is not directly related to
processing the entry-point. One such case is when there is not enough
memory (for example, due to all the other tasks being processed
simultaneously).

Previously, an `ENOMEM` error thrown on a worker process would propagate
to the master process, eventually causing ngcc to exit with an error.
Example failure: https://circleci.com/gh/angular/angular/682198

This commit improves handling of these low-memory situations by
detecting `ENOMEM` errors and killing the worker process, thus allowing
the master process to decide how to handle that. The master process will
put the task back into the tasks queue and continue processing tasks
with the rest of the worker processes (and thus with lower memory
pressure).

gkalpak committed:
refactor(ngcc): move `PathMappings` to separate file to avoid circular dependency

Now that `ngcc/src/ngcc_options` imports `FileWriter` type, there is a
circular dependency detected by the `ts-circular-deps:check` lint check:

```
ngcc/src/ngcc_options.ts
  → ngcc/src/writing/file_writer.ts
  → ngcc/src/packages/entry_point_bundle.ts
  → ngcc/src/ngcc_options.ts
```

This commit moves the `PathMappings` type (and related helpers) to a
separate file to avoid the circular dependency.

NOTE:
The circular dependency was only with taking types into account. There
was no circular dependency for the actual (JS) code.

gkalpak committed:
fixup! feat(ngcc): support reverting a file written by `FileWriter`

mary-poppins commented:
You can preview 95b94b0 at https://pr36626-95b94b0.ngbuilds.io/.
